### ISSUE TITLE
bundle install --deployment fails when gem source is specified twice, but Gemfile.lock is modified

### ISSUE THREAD

**Issue by tubbo** [bundler/bundler#4375]
Bundler will naturally raise an error after resolving dependencies to stop you from getting into dependency hell. Its stricter `--deployment` flag will flag any change to `Gemfile` which hasn't been reflected in `Gemfile.lock`.

On my local machine, running **v1.11.2**, the `bundle install` command ran fine and downloaded the proper gem. On my remote server, running **v1.10.3**, I get the following error on deploy:

```
00:06 bundler:install
      01 /opt/rbenv/bin/rbenv exec bundle install --path /var/www/foo/shared/bundle --without development test --deployment --quiet
(Backtrace restricted to imported tasks)
cap aborted!
SSHKit::Runner::ExecuteError: Exception while executing as webapp@10.100.12.124: bundle exit status: 16
bundle stdout: You are trying to install in deployment mode after changing
your Gemfile. Run `bundle install` elsewhere and add the
updated Gemfile.lock to version control.

You have changed in the Gemfile:
* foo-hearts from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-legacy_passwords from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-sitemaps from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-tag_manager from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
bundle stderr: Nothing written

SSHKit::Command::Failed: bundle exit status: 16
bundle stdout: You are trying to install in deployment mode after changing
your Gemfile. Run `bundle install` elsewhere and add the
updated Gemfile.lock to version control.

You have changed in the Gemfile:
* foo-hearts from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`


* foo-legacy_passwords from `no specified source` to `rubygems repository


 ¹Gemfile+                                                                                                                                                                                        buffers
https://gems.foo.com/, https://rubygems.org/`
* foo-sitemaps from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-tag_manager from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
bundle stderr: Nothing written

Tasks: TOP => deploy:updated => bundler:install
(See full trace by running task with --trace)
The deploy has failed with an error: Exception while executing as webapp@10.100.12.124: bundle exit status: 16
bundle stdout: You are trying to install in deployment mode after changing
your Gemfile. Run `bundle install` elsewhere and add the
updated Gemfile.lock to version control.

You have changed in the Gemfile:
* foo-hearts from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-legacy_passwords from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-sitemaps from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
* foo-tag_manager from `no specified source` to `rubygems repository
https://gems.foo.com/, https://rubygems.org/`
bundle stderr: Nothing written

** DEPLOY FAILED
** Refer to log/capistrano.log for details. Here are the last 20 lines:

DEBUG [907e1120] 	[31mYou are trying to install in deployment mode after changing
DEBUG [907e1120] 	your Gemfile. Run `bundle install` elsewhere and add the
DEBUG [ec9af203] 	https://gems.foo.com/, https://rubygems.org/`
DEBUG [907e1120] 	updated Gemfile.lock to version control.
DEBUG [ec9af203] 	* foo-sitemaps from `no specified source` to `rubygems repository
DEBUG [907e1120]
DEBUG [ec9af203] 	https://gems.foo.com/, https://rubygems.org/`
DEBUG [907e1120] 	You have changed in the Gemfile:
DEBUG [ec9af203] 	* foo-tag_manager from `no specified source` to `rubygems repository
DEBUG [907e1120] 	* foo-hearts from `no specified source` to `rubygems repository
DEBUG [ec9af203] 	https://gems.foo.com/, https://rubygems.org/`[0m
DEBUG [907e1120] 	https://gems.foo.com/, https://rubygems.org/`
DEBUG [907e1120] 	* foo-legacy_passwords from `no specified source` to `rubygems repository
DEBUG [907e1120] 	https://gems.foo.com/, https://rubygems.org/`
DEBUG [907e1120] 	* foo-sitemaps from `no specified source` to `rubygems repository
DEBUG [907e1120] 	https://gems.foo.com/, https://rubygems.org/`
DEBUG [907e1120] 	* foo-tag_manager from `no specified source` to `rubygems repository
DEBUG [907e1120] 	https://gems.foo.com/, https://rubygems.org/`[0m
DEBUG [27a5ebf5]
```

Fixed it by moving the only gem with a `git:` dependency outside of the `source` block, but I thought it might be useful to throw a warning or an error, in order to avoid future deployment problems.

**Comment by segiddins** [comment]
In this case, I think the easiest way for you to avoid the error would be to use the fixed version of bundler on your server as well as locally? (If I'm remembering correctly, this is a bug we fixed in 1.11)

**Comment by tubbo** [comment]
@segiddins Ah OK, so this is definitely is/was a bug. Thanks man!
